### PR TITLE
Config to exclude class prefix from serialize-messages, #24273

### PR DESCRIPTION
**Pull request by patriknw**
* Adding no-serialization-verification-needed-class-prefix, with "akka." included
  by default
* This is important now when we disable Java serialization by default.
  There will be many complaints about Akka internal classes that are not marked with
  NoSerializationVerificationNeeded, and we can't really sprinkle that everywhere.
* Not removing NoSerializationVerificationNeeded usage in Akka classes because that
  may break bin compat, even though it's needed any more.

Refs #24273

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/dungeon/Children.scala [hunk 1]
```diff
@@ -266,7 +267,8 @@ private[akka] trait Children { this: ActorCell =>
         props.args.forall(
           arg =>
             arg == null ||
-            arg.isInstanceOf[NoSerializationVerificationNeeded] || {
+            arg.isInstanceOf[NoSerializationVerificationNeeded] ||
+            settings.NoSerializationVerificationNeededClassPrefix.exists(arg.getClass.getName.startsWith) || {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/Children.scala]: here is one additional condition

File: akka-actor/src/main/scala/akka/actor/dungeon/Dispatch.scala [hunk 2]
```diff
@@ -169,14 +169,18 @@ private[akka] trait Dispatch { this: ActorCell =>
...
       case msg =>
-        val deserializedMsg = try {
-          serializeAndDeserializePayload(msg)
-        } catch {
-          case NonFatal(e) => throw SerializationCheckFailedException(msg, e)
-        }
-        envelope.message match {
-          case dl: DeadLetter => envelope.copy(message = dl.copy(message = deserializedMsg))
-          case _              => envelope.copy(message = deserializedMsg)
+        if (system.settings.NoSerializationVerificationNeededClassPrefix.exists(msg.getClass.getName.startsWith))
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/Dispatch.scala]: here is one additional condition

File: akka-actor/src/main/resources/reference.conf [hunk 3]
```diff
@@ -119,6 +119,10 @@ akka {
     # as marked with deploy.scope == LocalScope are exempt from verification.
     serialize-creators = off
 
+    # If serialize-messages or serialize-creators are enabled classes that starts with
+    # a prefix listed here are not verified.
+    no-serialization-verification-needed-class-prefix = ["akka."]
```
Comment (patriknw) [inline akka-actor/src/main/resources/reference.conf]: here is the new config

File: akka-docs/src/main/paradox/serialization.md [hunk 4]
```diff
@@ -302,3 +286,23 @@ It must still be possible to deserialize the events that were stored with the ol
...
+
+### Verification
+
+Normally, messages sent between local actors (i.e. same JVM) do not undergo serialization. For testing, sometimes, it may be desirable to force serialization on all messages (both remote and local). If you want to do this in order to verify that your messages are serializable you can enable the following config option:
+
+@@snip [SerializationDocSpec.scala](/akka-docs/src/test/scala/docs/serialization/SerializationDocSpec.scala) { #serialize-messages-config }
+
+Certain messages can be excluded from verification by extending the marker @scala[trait]@java[interface]
+`akka.actor.NoSerializationVerificationNeeded` or define a class name prefix in configuration
+`akka.actor.no-serialization-verification-needed-class-prefix`.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/serialization.md]: I moved section to end of page to deemphasize the feature, and added this paragraph

**Review by patriknw — COMMENTED** [review]
Most of the changes are removal/adjustment of test config.
Added a few comments to highlight the important changes.

File: akka-docs/src/main/paradox/serialization.md [hunk 5]
```diff
@@ -302,3 +286,23 @@ It must still be possible to deserialize the events that were stored with the ol
...
+
+If you want to verify that your `Props` are serializable you can enable the following config option:
+
+@@snip [SerializationDocSpec.scala](/akka-docs/src/test/scala/docs/serialization/SerializationDocSpec.scala) { #serialize-creators-config }
+
+@@@ warning
+
+We recommend having these config options turned on **only** when you're running tests. Turning these options on in production is pointless, as it would negatively impact the performance of local message passing without giving any gain.
+
+@@@
```
Comment (johanandren) [inline akka-docs/src/main/paradox/serialization.md]: I'd hoped we could not only de-emphasize but deprecate the feature now that we have the `SerializationTestKit`

**Review by johanandren — APPROVED** [review]

File: akka-docs/src/main/paradox/serialization.md (hunk 5 above)
Comment (patriknw) [reply akka-docs/src/main/paradox/serialization.md]: `SerializationTestKit` is only for Typed (so far) and it doesn't help detecting messages that don't have a serializer.

File: akka-actor-tests/src/test/scala/akka/event/LoggerSpec.scala [hunk 6]
```diff
@@ -66,6 +66,7 @@ object LoggerSpec {
         loggers = ["akka.event.LoggerSpec$$TestLogger1"]
         actor {
           serialize-messages = on
+          no-serialization-verification-needed-class-prefix = []
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/event/LoggerSpec.scala]: Do we explicitly want to configure it with an empty list here?

**Review by raboof — COMMENTED** [review]
Nice cleanup, one question but otherwise LGTM

**Comment by jrudolph** [comment]
Maybe we should discuss the utility of `serialize-message` at this occasion? Is it really useful? How would people use it in practice? It requires us and users to annotate all messages to distinguish between remote messages and local messages. If even we don't manage to use the feature as intended, why would our users?

**Comment by patriknw** [comment]
I don't like the feature, but just removing it without replacement (and we don't have bandwidth to implement some kind of serialization testkit now) seems harsh. Given that users complain all the time when we miss to mark messages with NoSerializationVerificationNeeded it must be a used feature.

**Comment by raboof** [comment]
```

**Review by raboof — APPROVED** [review]

**Pull request merged**